### PR TITLE
systemctl: systemctl show --property' needs verification of property

### LINKED ISSUE #2295: 'systemctl show --property' needs verification of property requested and return correct exit code
Description of problem:

systemctl --property doesn't validate if a requested property is valid or not, and always returns with an exit code of 0, regardless of whether the requested property exists or not.

How reproducible:

This works fine:

```
# systemctl show --property Id default.target
Id=multi-user.target
```

But put in a non-existing property:

```
# systemctl show --property Id default.targets
Id=default.targets.service
# systemctl show --property Id default.targets; echo $?
Id=default.targets.service
0
#
[root@shou18lkvm8 ~]# systemctl show --property Id this.is.rubbish; echo $?
Id=this.is.rubbish.service
0
```

### PR DESCRIPTION
**Pull request by ssahani**
fixes #2295

### CODE REVIEW

File: src/systemctl/systemctl.c [hunk 1]
```diff
@@ -4525,6 +4536,19 @@ static int show_one(
         if (r < 0)
                 return log_error_errno(r, "Failed to get properties: %s", bus_error_message(&error, r));
 
+        r = bus_message_map_all_properties(reply, property_map, &property_info);
+        if (r < 0)
+                return log_error_errno(r, "Failed to map properties: %s", bus_error_message(&error, r));
+
+        if (!isempty(property_info.load_state) && !isempty(property_info.active_state)) {
+                if (streq(property_info.load_state, "not-found") && streq(property_info.active_state, "inactive"))
```
Comment (keszybz) [inline src/systemctl/systemctl.c]: This outer condition can only be true if the inner condition is true. So you can simplify this to

``` C
if (streq_ptr(property_info.load_state, "not-found") &&
    streq_ptr(property_info.active_state, "inactive"))
```

File: src/systemctl/systemctl.c [hunk 2]
```diff
@@ -4831,6 +4855,9 @@ static int show(int argc, char *argv[], void *userdata) {
                                         return r;
                                 else if (r > 0 && ret == 0)
                                         ret = r;
+
+                                if (r == 3)
+                                        printf("Can't display property. Unit %s does not exists.", *name);
```
Comment (keszybz) [inline src/systemctl/systemctl.c]: `log_notice` or so.

"Does not exist." (no 's').

File: src/systemctl/systemctl.c [hunk 3]
```diff
@@ -4525,6 +4536,19 @@ static int show_one(
         if (r < 0)
                 return log_error_errno(r, "Failed to get properties: %s", bus_error_message(&error, r));
 
+        r = bus_message_map_all_properties(reply, property_map, &property_info);
+        if (r < 0)
+                return log_error_errno(r, "Failed to map properties: %s", bus_error_message(&error, r));
+
+        if (!isempty(property_info.load_state) && !isempty(property_info.active_state)) {
+                if (streq(property_info.load_state, "not-found") && streq(property_info.active_state, "inactive"))
+                        return 3;
```
Comment (keszybz) [inline src/systemctl/systemctl.c]: Why "3"? I think -ENOENT or -ESRCH would be better.

File: src/systemctl/systemctl.c [hunk 4]
```diff
@@ -4488,6 +4488,11 @@ static int print_property(const char *name, sd_bus_message *m, const char *conte
         return 0;
 }
 
+struct property_info {
+                char *load_state;
+                char *active_state;
+};
```
Comment (keszybz) [inline src/systemctl/systemctl.c]: This structure definition should be folded into the variable definition below. And `const char *` would be better:

```
struct property_info {
    const char *load_state, *active_state;
} property_info = {};
```

**Comment by keszybz** [comment]
I think it's a nifty way to solve the issue. Looks good apart from some style issues.

Also, the commit message doesn't really do it justice: it's about checking that the unit does not exist, and not verifying properties.

File: src/systemctl/systemctl.c (hunk 2 above)
Comment (ssahani) [inline src/systemctl/systemctl.c]: We need to display to the user same as if a unit does not exist at all.

File: src/systemctl/systemctl.c (hunk 3 above)
Comment (ssahani) [inline src/systemctl/systemctl.c]: To be compatible with init scripts.
http://refspecs.linuxbase.org/LSB_3.1.0/LSB-Core-generic/LSB-Core-generic/iniscrptact.html

I think 4 is better BTW.

File: src/systemctl/systemctl.c [hunk 5]
```diff
@@ -4523,6 +4523,10 @@ static int print_property(const char *name, sd_bus_message *m, const char *conte
         return 0;
 }
 
+struct property_info {
+        const char *load_state, *active_state;
+} property_info = {};
```
Comment (keszybz) [inline src/systemctl/systemctl.c]: This should not be global table. It should be defined in the function.

File: src/systemctl/systemctl.c [hunk 6]
```diff
@@ -4562,6 +4571,17 @@ static int show_one(
         if (r < 0)
                 return log_error_errno(r, "Failed to get properties: %s", bus_error_message(&error, r));
 
+        r = bus_message_map_all_properties(reply, property_map, &property_info);
+        if (r < 0)
+                return log_error_errno(r, "Failed to map properties: %s", bus_error_message(&error, r));
+
+        if (streq_ptr(property_info.load_state, "not-found") && streq_ptr(property_info.active_state, "inactive"))
+                return 4;
```
Comment (keszybz) [inline src/systemctl/systemctl.c]: <strike>You changed the 3 to 4, but you still didn't answer the question, why this cannot be -EXISTS or -ESRCH instead?</strike> Sorry, I missed your comment.

File: src/systemctl/systemctl.c (hunk 5 above)
Comment (ssahani) [inline src/systemctl/systemctl.c]: If something else requires it then  we don't again have to make it local->global.

**Comment by keszybz** [comment]
OK, so I think this is good to go in, except that the table definition that should not be global.

File: src/systemctl/systemctl.c (hunk 6 above)
Comment (ssahani) [inline src/systemctl/systemctl.c]: Ooops I  was about to write never mind thanks

File: src/systemctl/systemctl.c (hunk 5 above)
Comment (keszybz) [inline src/systemctl/systemctl.c]: No, it's used locally in the function. There is no reason for this to be global.

Comment (ssahani) [inline src/systemctl/systemctl.c]: for now yes . This is not the point actually .

Comment (keszybz) [inline src/systemctl/systemctl.c]: No, the variable can never be global. The structure definition could be, but it's nicer to define it scoped to the function too.

Comment (ssahani) [inline src/systemctl/systemctl.c]: <p><strike>I am making the change but I am not sure.</p></strike> Yes the structure definition let's make it global not the variable.

Comment (ssahani) [inline src/systemctl/systemctl.c]: sorry I dont's know how to cross the comments. I am editing it . Let's keep the definition global and variables for this local.

**Comment by ssahani** [comment]
We can always make it global later sorry for confusion.

File: src/systemctl/systemctl.c [hunk 7]
```diff
@@ -4868,6 +4887,9 @@ static int show(int argc, char *argv[], void *userdata) {
                                         return r;
                                 else if (r > 0 && ret == 0)
                                         ret = r;
+
+                                if (r == 4)
+                                        printf("Can't display property %s. Unit %s does not exist.", *patterns, *name);
```
Comment (poettering) [inline src/systemctl/systemctl.c]: errors should be printed with log_error() or such like, not with printf(), especially not to stdout.

**Comment by poettering** [comment]
looks pretty ok, just one more fix.

**Comment by ssahani** [comment]
force pushed . Probably I should also post 
in the TODO list
- Rework systemctl's GetAll property parsing to use the generic bus_map_all_properties() API

**Comment by poettering** [comment]
yeah, we should probably add something like that to the TODO list.

Patch looks good now. Let's wait for the CI and merge it.

**Comment by ssahani** [comment]
I have updated according to new init script error codes (enum).

**Pull request merged**